File: src/LanAudioRelay.Core/Networking/ControlClient.cs
using System.Net;
using System.Net.Sockets;
using LanAudioRelay.Core.Audio;
using LanAudioRelay.Core.Protocol;

namespace LanAudioRelay.Core.Networking;

public sealed class ControlClient
{
    public async Task<PairingSession> ConnectAndPairAsync(
        string host,
        string pairingCode,
        int controlPort = AudioConstants.ControlPort,
        CancellationToken cancellationToken = default)
    {
        using var tcpClient = new TcpClient();
        await tcpClient.ConnectAsync(host, controlPort, cancellationToken).ConfigureAwait(false);

        await using var stream = tcpClient.GetStream();
        using var connection = new JsonLineConnection(stream);

        var clientNonce = Pairing.CreateNonce();
        await connection.SendAsync(
            new HelloRequest(
                "hello",
                AudioConstants.ProtocolVersion,
                Environment.MachineName,
                clientNonce),
            cancellationToken).ConfigureAwait(false);

        var challenge = await connection.ReceiveAsync<ChallengeResponse>(cancellationToken).ConfigureAwait(false);
        if (challenge.Type != "challenge" || challenge.ProtocolVersion != AudioConstants.ProtocolVersion)
        {
            throw new InvalidDataException("The receiver returned an unsupported control challenge.");
        }

        var proof = Pairing.CreateProof(pairingCode, clientNonce, challenge.ServerNonce);
        await connection.SendAsync(new PairRequest("pair", proof), cancellationToken).ConfigureAwait(false);

        var result = await connection.ReceiveAsync<PairResult>(cancellationToken).ConfigureAwait(false);
        if (!result.Accepted)
        {
            throw new UnauthorizedAccessException(result.ErrorMessage ?? "The receiver rejected the pairing code.");
        }

        return new PairingSession(result.SessionId, host, result.MediaPort, result.AudioSettings);
    }
}

File: mac/LanAudioRelayMac/Sources/LanAudioRelayMacCore/ControlClient.swift
import Foundation

public final class ControlClient {
    public init() {}

    public func connectAndPair(
        host: String,
        pairingCode: String,
        controlPort: UInt16 = AudioConstants.controlPort
    ) async throws -> PairingSession {
        let connection = TCPLineConnection(host: host, port: controlPort)
        try await connection.start()
        defer { connection.cancel() }

        let clientNonce = Pairing.createNonce()
        try await connection.send(HelloRequest(
            clientName: Host.current().localizedName ?? "Mac",
            clientNonce: clientNonce
        ))

        let challenge = try await connection.receive(ChallengeResponse.self)
        guard challenge.type == "challenge",
              challenge.protocolVersion == AudioConstants.protocolVersion
        else {
            throw ProtocolError.unsupportedProtocol
        }

        let proof = Pairing.createProof(
            code: pairingCode,
            clientNonce: clientNonce,
            serverNonce: challenge.serverNonce
        )
        try await connection.send(PairRequest(proof: proof))

        let result = try await connection.receive(PairResult.self)
        guard result.accepted else {
            throw ProtocolError.pairingRejected(result.errorMessage ?? "Pairing rejected")
        }

        return PairingSession(
            sessionId: result.sessionId,
            receiverHost: host,
            mediaPort: result.mediaPort,
            audioSettings: result.audioSettings
        )
    }
}
